### ISSUE TITLE
A handle for transactions

### ISSUE THREAD

**Issue by F21** [arangodb/arangodb#471]
Jan has provided a great reply to how transactions will work in my discussion [here](https://groups.google.com/forum/?fromgroups#!topic/arangodb/M7TaymcdQYU).

I have thought of an idea that can perhaps allow us to have commits and rollbacks. Since I am well versed with the REST interface, I will show my example using REST calls:

We post a transaction:

```
POST /transactions

{
  //statement 1
  //statement 2
  ...
  "handle": true //this asks the server to return a handle. If this is false or not present, the server tries to commit the batch immediately.
}
```

We get this response:

```
{
  "handle": /transactions/handles/2905300qijkaofdgjapoisjoapsgodapsg //a handle with a random id
}
```

To commit:

```
POST /transactions/handles/2905300qijkaofdgjapoisjoapsgodapsg/commit
```

To rollback:

```
POST /transactions/handles/2905300qijkaofdgjapoisjoapsgodapsg/rollback
```

This idea is taken from the RESTful Web Wervices Cookbook: http://books.google.com.au/books?id=LDuzpQlVuG4C&lpg=PP1&pg=PA213#v=onepage&q&f=false

What do you guys think about this? :smile:

**Comment by solher** [comment]
Hi,

I'm a simple ArangoDB user and I'm working with Go. To me, the way the transactions are currently done feels really weird when you are not using the javascript client.

I would like to do something like that, when creating a node for a user:
```
LET newNode = (INSERT {} IN nodes RETURN NEW)
INSERT {"_from": "users/7859415327", "_to": newNode._id} IN links
RETURN newNode
```

I understand I can't do that so I have to use transactions to have an atomic operation.

The problem is I just can't get the result of the first insert to use it in the second one without writing a whole javascript function. Which feels really weird and clumsy because you end up having a bunch of javascript strings everywhere in your Go code. So I've given up and I don't use transactions anymore. 

I would personally feel more comfortable with a traditional BEGIN COMMIT ROLLBACK schema.

That's just my humble observation.
Thanks for your amazing work.

**Comment by jsteemann** [comment]
I think supporting arbitrary crud operations inside a "traditional" `begin`...`commit` schema is the most intuitive way to do it, as most users are accustomed to this. Additionally, it will also be the most flexible solution. But it requires intense API and internal changes, because transactions would need to remain open across multiple HTTP requests etc.

By the way, if the type of operations to perform in a transaction are always the same or mostly identical, it will also be possible to code the transaction in JavaScript, and deploy it as a Foxx action that your application can call. This way your application can send only the actual data (and no JavaScript code) to the server, the server will carry out the transaction in the Foxx route using the request data, and can send back the results to your application in a format that is sensible for it.
Does that make sense or do you need further examples?

**Comment by solher** [comment]
It makes perfect sense. But it feels like moving the problem.
It means that you are forced to put the actual AQL queries partly in the client code and partly in the server side and it is a problem for maintainability.

But yeah, I understand that ArangoDB really encourages the full use of Foxx and the use of the DB more as a REST server than a traditional database.
I'm just not sure it is a good point for maintainability to have a microservice oriented architecture where the logic is half in the service and half written in another language on another server. It just feels wrong.

And it's not about having a transaction open across multiple HTTP requests. It would be perfectly fine for me to have transactions part of AQL. Something like:

```
BEGIN
     RUN newNode = (INSERT {} IN nodes RETURN NEW)

     RUN newLink = (INSERT {"_from": "users/7859415327", "_to": newNode._id} IN links RETURN NEW)

     RETURN newNode
COMMIT
```

But I guess it would add more problems than solutions haha.

**Comment by clintwood** [comment]
@jsteemann, it's been a while since this issue was raised - are there any (new) plans regarding transactions?

I agree with @solher about this 'not [being about] having a transaction open across multiple HTTP requests' - I would strongly discourage handles across a HTTP/REST API calls, this would be a pattern which would introduce unnecessary complexity both client and server side (IMO transaction timeouts, as suggested earlier (or similar 'hacks'), really aren't a solution here either).

(FWIW, my current solution/hack is to generate javascript that wraps several `db._query(...)` statements which execute via the `transaction` HTTP API which is not ideal).

**There seems to be an obvious solution:** Since AQL queries run in the context of a transaction it seems that allowing more than a single 'data modification query' within an AQL query and/or allowing a 'data modification query' within a AQL subquery would solve this issue completely.  Has this been considered and if so is there any reason that it is not being implemented?

**Comment by solher** [comment]
Obviously, that would solve the problem. I would be interested to know the reasons too.

Another thing I realized since then is that the transaction HTTP API doesn't support cursors. So you just can't retrieve multiple batches when the result of the transaction is huge, which can be a problem sometimes.

Using a regular query would solve the problem too...

**Comment by jsteemann** [comment]
@clintwood : using an AQL query that can modify multiple collections at once is the obvious solution, and I think this would cover many cases already (though there will be some rarer cases in which you really want a handle that the client can use across multiple requests).
I had started a local branch for multi-collection modification queries a while ago, but did not find time to continue working on it. It's probably outdated now, but I can try to reactivate that and check how much work it would be to finish the feature.

**Comment by clintwood** [comment]
@jsteemann, that would be really awesome - I'm really surprised there's not more call for 'multi-collection modifications'.  I'd love to contribute on this but I'm just totally oversubscribed time-wise! 

A critical point to consider if you are looking into this would be a way to capture the output of the data modification statement into a local variable via LET so that it could be used in subsequent queries, something like the [code shown in this message](https://github.com/arangodb/arangodb/issues/471#issuecomment-141075650) but then that's kinda obvious :)

Thanks for looking at this!

**Comment by fceller** [comment]
@clintwood could you contact the ArangoDB team at 

hackers (at) arangodb.org

Thanks

**Comment by jsteemann** [comment]
I pushed the current state into the `aql-multi-modify` branch, which is based on current devel.
The branch has an half-way working, fully untested version of multi-collection modification queries. I can't guarantee I can continue working on this soon, but at least the progress can be tracked now and the work is not lost.

**Comment by clintwood** [comment]
@jsteemann, great news, will take a look!

**Comment by clintwood** [comment]
@jsteemann, Firstly this is really great!
I managed to build the `aql-multi-modify` branch and briefly played around with some queries.  In the most part most of what I need works.  Not being that familiar with the codebase I wasn't sure exactly what should/shouldn't work.  There were a few errors that pointed to issues related to running multiple statements but the primary one was "access after data-modification (while optimizing ast)".  This occurs when trying to query/access a collection just after it has been modified (see commented code below).

Other than that this is a really excellent start (thanks) and IMO opens up a so many possibilities for AQL!

```
// requires: odd, even, all collections

// generate some test data
LET data = (
  FOR i IN 0..10
  RETURN {id: i, value: CONCAT('value - ', i)}
)

// insert odds
LET newOdds = (
  FOR d IN data
  FILTER (d.id % 2) // odd
  INSERT d INTO odd
  RETURN NEW
)

// insert evens
LET newEvens = (
  FOR d IN data
  FILTER !(d.id % 2) // even
  INSERT d INTO even
  RETURN NEW
)

// insert UNION of odd & even
LET unionall = (
  FOR a IN UNION(newOdds, newEvens)
  INSERT a INTO all
  RETURN NEW
)

// LET all = (
//   FOR a IN all
//   SORT a.id
//   RETURN {id: a.id, type: a.type}
// )

return {
  newOdds,
  newEvens,
  unionall
  // all
}
```

**Comment by clintwood** [comment]
And this works (@solher) :

```
// add new node
LET newNode = (INSERT {} IN nodes RETURN NEW)
// link user to new node
INSERT {_from: 'users/1', _to: newNode[0]._id} INTO links
RETURN newNode
```

**Comment by jsteemann** [comment]
@clintwood: the idea was that a collection can be modified in a query, but that the collection must not be used after the modification statement.

For example: `FOR doc IN collection REMOVE doc IN collection` will work because `collection` is not mentioned in the query after the remove statement. The following (nonsense) queries however will fail with the above error, because one of the modified collections is accessed after being modified:

* `FOR doc IN collection REMOVE doc.foo IN collection REMOVE doc.bar IN collection`
* `FOR doc IN collection REMOVE doc.foo IN collection FOR doc2 IN collection RETURN doc2`
* `FOR doc IN collection REMOVE doc.foo IN collection RETURN (FOR doc2 IN collection RETURN doc2)`

Additionally, any operation that may access data from a modified collection cannot be used after a modification operation either. This includes AQL functions that access collection data such as `DOCUMENT`, `EDGES`, `NEAR`, `WITHIN`, `GRAPH_...`.

A particular problem may be that there cannot be two inserts or remove operations on the same collection in one query. However this is doable by rewriting the query slightly to use just one remove operation but call it in a loop.

* Original (won't work): `FOR doc IN collection REMOVE doc.foo IN collection REMOVE doc.bar IN collection`
* Workaround (should work): `FOR doc IN collection FOR what IN [ doc.foo, doc.bar] REMOVE what IN collection`

**Comment by clintwood** [comment]
@jsteemann, definitely some caveats here but I think with a bit more thought/tweaking this addition would make AQL extremely powerful in a transaction environment!  Would be great to be able to use the collections after modification but I assume that would get complex - but probably worth looking into!

**Comment by clintwood** [comment]
@jsteemann, @fceller, this is really useful as it is now!  I realise it's basically untested and possibly conceptually not complete - so the question is when will this (given that tests pass) be pushed to production.  The reason I ask is, should this land in production before end of Jan 2016 I would probably end up maintaining a private fork so that my current project can be built using this functionality now.

**Comment by jsteemann** [comment]
@clintwood : our goal is to add the feature to the upcoming ArangoDB version (i.e. 2.8). 
However, it's not yet in the devel branch. We need to write many more tests before we can pull it. If any one of the to-be-written tests reveals a conceptual problem issue with the feature, then we may have to rethink the situation. 
Long story short: if everything works well, then I think we will pull it into 2.8, but if not, then we can't. 2.8 is not too far ahead (plan is to release it in December I think), so we hopefully have certainty soon.

**Comment by clintwood** [comment]
@jsteemann, great, thanks for the feedback!  I've merged `aql-multi-modify` with a very recent `devel` branch to use for dev - It's really nice to have that feature!  Thanks for looking into it - holding thumbs for inclusion in 2.8!

**Comment by jsteemann** [comment]
@clintwood: by the way, the `aql-multi-modify` branch has been merged into the `devel` and `2.8` branches recently. Bugfixes will only happen there. The feature will also become part of a 2.8 release.

**Comment by clintwood** [comment]
@jsteemann, Thank you sir!!!  Will definitely give that a whirl!

**Comment by graetzer** [comment]
https://www.arangodb.com/docs/stable/http/transaction-stream-transaction.html
